RecoLocalMuon/GEMSegment/plugins/ME0SegAlgoMM.cc

```diff
@@ -0,0 +1,351 @@
+    eta_new.push_back(pos_new.eta());
+  }  
+  for(size_t iRH_old = 0;iRH_old<oldChain.size();++iRH_old){
+    GlobalPoint pos_old = theEnsemble.first->toGlobal(oldChain[iRH_old]->localPosition());
+    layer_old.push_back(oldChain[iRH_old]->me0Id().layer());
+    phi_old.push_back(pos_old.phi());
+    eta_old.push_back(pos_old.eta());
+  }
+
+  for(size_t jRH_new = 0; jRH_new<phi_new.size(); ++jRH_new){
+    for(size_t jRH_old = 0; jRH_old<phi_old.size(); ++jRH_old){
+
+      // to be chained, two hits need to be in neighbouring layers...
+      // or better allow few missing layers (upto 3 to avoid inefficiencies);
+      // however we'll not make an angle correction because it
+      // worsen the situation in some of the "regular" cases 
+      // (not making the correction means that the conditions for
+      // forming a cluster are different if we have missing layers -
+      // this could affect events at the boundaries ) 
+
+      // to be chained, two hits need also to be "close" in phi and eta
+      bool phiRequirementOK = reco::deltaPhi(phi_new[jRH_new],phi_old[jRH_old]) < dPhiChainBoxMax;
+      bool etaRequirementOK = fabs(eta_new[jRH_new]-eta_old[jRH_old]) < dEtaChainBoxMax;
+      // and the difference in layer index should be < (nlayers-1)
+      bool layerRequirementOK = abs(layer_new[jRH_new]-layer_old[jRH_old]) < (theEnsemble.first->id().nlayers()-1);
+
+      if(layerRequirementOK && phiRequirementOK && etaRequirementOK){
+        return true;
+      }
+    }
+  } 
+  return false;
+}
+
+
+
+
+
+std::vector<ME0Segment> ME0SegAlgoMM::buildSegments(const EnsembleHitContainer& rechits) {
+  std::vector<ME0Segment> me0segs;
+
+  proto_segment.clear();
+  // select hits from the ensemble and sort it 
+  for (auto rh=rechits.begin(); rh!=rechits.end();rh++){
+    proto_segment.push_back(*rh);
+  }
+  if (proto_segment.size() < minHitsPerSegment){
+    return me0segs;
+  }
+
+  // The actual fit on all hits of the vector of the selected Tracking RecHits:
+  sfit_ = std::unique_ptr<ME0SegFit>(new ME0SegFit(theEnsemble.second, rechits));
+  sfit_->fit();
+  edm::LogVerbatim("ME0SegAlgoMM") << "[ME0SegAlgoMM::buildSegments] ME0Segment fit done";
+
+  // obtain all information necessary to make the segment:
+  LocalPoint protoIntercept      = sfit_->intercept();
+  LocalVector protoDirection     = sfit_->localdir();
+  AlgebraicSymMatrix protoErrors = sfit_->covarianceMatrix(); 
+  double protoChi2               = sfit_->chi2();
+
+  // Calculate the central value and uncertainty of the segment time
+  float averageTime=0.;
+  for (auto rh=rechits.begin(); rh!=rechits.end(); ++rh){
+    averageTime += (*rh)->tof();                                          
+  }
+  if(rechits.size() != 0) averageTime=averageTime/(rechits.size());
+  float timeUncrt=0.;
+  for (auto rh=rechits.begin(); rh!=rechits.end(); ++rh){
+    timeUncrt += pow((*rh)->tof()-averageTime,2);
+  }
+  if(rechits.size() != 0) timeUncrt=timeUncrt/(rechits.size());
```

Comment:
noted, but does not matter, because the segment is only made when there are 2 or more hits.